Add a design doc for the SIMD (HECO batching) optimizations

AlexanderViand-Intel opened the pull request:
I still need to add a more detailed description of the main `-insert-rotate` pass, but this should already give a bit of an overview over the whole batching pipeline and I'd appreciate any feedback/questions. I realized I've gotten too used to "academic paper" language, so it doesn't sound as "documentation-y" as I wanted 😅 

Oh, and I noticed the `-apply-folders` pass wasn't registered in `heir-opt` while playing around with the pipeline to verify it works as  described, but I guess that should be a separate bug-fix PR.

EDIT: I guess we can either keep this as a draft until Jeremy is back, or merge it once it's completed/polished a bit and add a todo/issue for him to do a pass on it.

j2kun commented on docs/content/en/docs/Design/simd.md:
consistency (sp)

j2kun commented on docs/content/en/docs/Design/simd.md:
it's -> its

j2kun commented on docs/content/en/docs/Design/simd.md:
Is it worth explaining what those canoncalization patterns are?

AlexanderViand-Intel replied:
I added a brief "tensor text canonicalizations" section to the document and referenced it from here, but I feel like this information really out to go to https://heir.dev/docs/dialects/tensorext/, I'm just not sure where I'd have to put it for it to appear there.

asraa replied:
I agree it feels like it should belong to canonicalization in that dialect. I bet we can just add a README.md to each Dialect subfolder and then append that to the markdown generated https://github.com/google/heir/blob/15e510cf97874bac4f7901bc0fe5556e01bc7294/.github/workflows/copy_tblgen_files.sh#L64

right below the header: https://heir.dev/docs/dialects/tensorext/

Maybe the secret design section should actually go in like that too, and maybe each dialect should optionally have  a Design/MoreInfo subpage. We can discuss on an issue

j2kun reviewed (approved):
Looks like a good start!

j2kun commented on docs/content/en/docs/Design/simd.md:
File an issue and replace this with `TODO(#123):` where 123 is replaced with the issue number

AlexanderViand-Intel replied:
I was originally planning to finish writing this up, but I think I'll take the easy way out and do just that ;)

AlexanderViand-Intel commented:
Thanks for the review @j2kun! I added a description of the `tensor_ext` canonicalization patterns, fixed the typos you spotted (and an embarrassingly high number of additional typos) and created #721 for the remaining TODO.

Oh, and I should probably mention the original "Design Doc" issue #629 so this PR gets linked from there, too.

asraa commented on docs/content/en/docs/Design/simd.md:
Useful warning! Next HEIR meeting I may pick brains at this. I started writing a packing pass that promotes arbitrary tensors into tensors with a desired power of two number of slots with padding/replication, but I'm a little stuck on multi-dimensional tensors (which is probably necessary for matrix multiplication). It seems to me slightly chicken-and-eggy to write a packing pass and then find optimal rotations which depend on the packing. But like you mention below, we will always be relying on heuristics, and so some kind of heuristic will probably be necessary (and maybe we allow for handwritten attributes or options that can indicate a better approach)

asraa commented on docs/content/en/docs/Design/simd.md:
One question I have that I am happy to add an answer to when I understand is when you want to use one or the either. (Or is it even possible to use straight-line-vectorize to tensor-ify operations and then heir-simd-vectorize after? probably doesn't do anything useful)

asraa reviewed (approved):
Thanks!
